lc0/src/neural/network_cudnn.cu

```diff
@@ -0,0 +1,836 @@
+            FCVal2->loadWeights(&weights.ip2_val_w[0], &weights.ip2_val_b[0]);
+            network.push_back(FCVal2);
+        }
+        valueOut = getLastLayer();
+
+        // 2. allocate GPU memory for running the network
+        //    - three buffers of max size are enough (one to hold input, second to hold output and third to hold skip connection's input)
+        resiLast->setBatchSize(MAX_BATCH_SIZE);
+        size_t maxSize = resiLast->getOutputSize();
+        for (int i = 0; i < 3; i++)
+        {
+            cudaMalloc(&tensorMem[i], maxSize);
+            cudaMemset(tensorMem[i], 0, maxSize);
+        }
+
+        //printf("Allocated %d bytes of GPU memory to run the network\n", 3 * maxSize);
+
+        // 3. allocate scratch space (used internally by cudnn to run convolutions)
+        cudaMalloc(&scratchMem, CUDA_SCRATCH_SIZE);
+
+    }
+
+    void forwardEval(float *input, float *op_pol, float *op_val, int batchSize) const
+    {
+        // printf(" ..%d.. ", batchSize);
+
+        g_lock.lock();
+
+        BaseLayer::setBatchSize(batchSize);
+
+        // copy data from CPU memory to GPU memory
+        cudaMemcpyAsync(tensorMem[0], &input[0], 
+                        BaseLayer::getN() * kInputPlanes * network[0]->getH() * network[0]->getW() * sizeof(float), 
+                        cudaMemcpyHostToDevice);
+
+        int l = 0;
+        // input
+        network[l++]->eval(tensorMem[2], tensorMem[0], NULL, scratchMem);  // input conv
+
+        // residual block
+        for (int block = 0; block < numBlocks; block++)
+        {
+            network[l++]->eval(tensorMem[0], tensorMem[2], NULL, scratchMem);           // conv1
+            network[l++]->eval(tensorMem[2], tensorMem[0], tensorMem[2], scratchMem);   // conv2
+        }
+
+        // policy head
+        network[l++]->eval(tensorMem[0], tensorMem[2], NULL, scratchMem);    // pol conv
+        network[l++]->eval(tensorMem[1], tensorMem[0], NULL, scratchMem);    // pol BN
+        network[l++]->eval(tensorMem[0], tensorMem[1], NULL, scratchMem);    // pol FC       
+        network[l++]->eval(tensorMem[1], tensorMem[0], NULL, scratchMem);    // pol softmax  // POLICY
+
+        // value head
+        network[l++]->eval(tensorMem[0], tensorMem[2], NULL, scratchMem);    // value conv
+        network[l++]->eval(tensorMem[2], tensorMem[0], NULL, scratchMem);    // value BN
+        network[l++]->eval(tensorMem[0], tensorMem[2], NULL, scratchMem);    // value FC1
+        network[l++]->eval(tensorMem[2], tensorMem[0], NULL, scratchMem);    // value FC2    // VALUE
+
+        // copy results back to CPU memory
+        cudaMemcpyAsync(&op_pol[0], tensorMem[1], policyOut->getOutputSize(), cudaMemcpyDeviceToHost);
+        cudaError_t status = cudaMemcpy(&op_val[0], tensorMem[2], valueOut->getOutputSize(), cudaMemcpyDeviceToHost);
+
+        if (status != cudaSuccess)
+        {
+            printf("\nSome error running cuda based eval!\n");
```

Comment:
Consider changing to either assert of Exception.

Reply:
done.